cachi2/core/package_managers/pip.py

```diff
@@ -1788,27 +1915,31 @@ def _process_package_distributions(
                 ),
             )
     else:
-        log.warning("No source distributions found for package %s==%s", name, version)
-        best_sdist = None
+        log.warning("No sdist found for package %s==%s", name, version)
 
         if len(wheels) == 0:
             if allow_binary:
-                solution = "Please check that the package exists on PyPI or that the name and version are correct.\n"
+                solution = (
+                    "Please check that the package exists on PyPI or that the name"
+                    " and version are correct.\n"
+                )
                 docs = None
             else:
                 solution = (
-                    "It seems that this version does not exist or isn't published as a source distribution.\n"
-                    "Try to specify the dependency directly via a URL instead, for example, the tarball for a GitHub release."
+                    "It seems that this version does not exist or isn't published as an"
+                    " sdist.\n"
+                    "Try to specify the dependency directly via a URL instead, for example,"
+                    " the tarball for a GitHub release.\n"
+                    "Alternatively, allow the use of wheels."
                 )
                 docs = PIP_NO_SDIST_DOC
-
             raise PackageRejected(
                 f"No distributions found for package {name}=={version}",
                 solution=solution,
                 docs=docs,
             )
 
-    return best_sdist, wheels
+    return best_sdist + wheels if best_sdist else wheels
```

Comment:
I think what me and Erik were suggesting, is that `best_sdist` should not be a list variable. But you can return list from this function.

Reply:
I want to return a single list.

Reply:
@slimreaper35 I think the problem here is that `best_sdist` can be None in which case you'd have an extra element in the list unless you keep the conditional @ben-alkov  has in there. Can't we have an `artifacts` (`dpis` or `packages` or `ret`) variable and do:
```
ret = []
if sdists: 
    best_sdist = max()
    if best_sdist.is_yanked:
        ...
    ret.append(best_sdist)

...

ret.extend(wheels)
return artifacts
```

That way you wouldn't need the conditional in the return statement. Note that I know you were asked to drop the `artifacts` variable in comment https://github.com/containerbuildsystem/cachi2/pull/467#discussion_r1514082878, but I think we can simplify here easily.

Reply:
`ret.append(best_sdist)` can still be None, so we still need to filter it out

Reply:
how ?

Reply:
> `ret.append(best_sdist)` can still be None, so we still need to filter it out

Once you have a non-empty `sdists` list (we just checked it) then you can't end up with `best_sdist` being None coming from `max` or can you?